src/ash/bltin/echo.1
.\"	$NetBSD: echo.1,v 1.13 2003/08/07 09:05:40 agc Exp $
.\"
.\" Copyright (c) 1991, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" This code is derived from software contributed to Berkeley by
.\" Kenneth Almquist.
.\" Copyright 1989 by Kenneth Almquist
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"	@(#)echo.1	8.1 (Berkeley) 5/31/93
.\"
.Dd May 31, 1993
.Dt ECHO 1
.Os
.Sh NAME
.Nm echo
.Nd produce message in a shell script
.Sh SYNOPSIS
.Nm
.Op Fl n | Fl e
.Ar args ...
.Sh DESCRIPTION
.Nm
prints its arguments on the standard output, separated by spaces.
Unless the
.Fl n
option is present, a newline is output following the arguments.
The
.Fl e
option causes
.Nm
to treat the escape sequences specially, as described in the following
paragraph.
The
.Fl e
option is the default, and is provided solely for compatibility with
other systems.
Only one of the options
.Fl n
and
.Fl e
may be given.
.Pp
If any of the following sequences of characters is encountered during
output, the sequence is not output.  Instead, the specified action is
performed:
.Bl -tag -width indent
.It Li \eb
A backspace character is output.
.It Li \ec
Subsequent output is suppressed.  This is normally used at the end of the
last argument to suppress the trailing newline that
.Nm
would otherwise output.
.It Li \ef
Output a form feed.
.It Li \en
Output a newline character.
.It Li \er
Output a carriage return.
.It Li \et
Output a (horizontal) tab character.
.It Li \ev
Output a vertical tab.
.It Li \e0 Ns Ar digits
Output the character whose value is given by zero to three digits.
If there are zero digits, a nul character is output.
.It Li \e\e
Output a backslash.
.El
.Sh HINTS
Remember that backslash is special to the shell and needs to be escaped.
To output a message to standard error, say
.Pp
.D1  echo message \*[Gt]\*[Am]2
.Sh BUGS
The octal character escape mechanism
.Pq Li \e0 Ns Ar digits
differs from the
C language mechanism.
.Pp
There is no way to force
.Nm
to treat its arguments literally, rather than interpreting them as
options and escape sequences.


ChangeLog
/* $Id: ChangeLog 644 2006-12-03 05:06:18Z knut.osmundsen@oracle.com $ */

2006-12-02:
    - kmk:
        o Fixed bugs dealing with the order SDK properties was applied to sources.

2006-12-01:
    - kBuild:
        o Added support for compiling resource files.
        o Fixed bug causing object files take from the SOURCES to be cleaned up.

2006-11-25:
    - kmk, kBuild:
        o Added mv as builtin command.
        o Added cat as builtin command.
        o Added GNU sed version 4.1.5 and ported it to MSC.
        o Added rmdir as builtin and external command.
        o Made mkdir not modify the argument strings.
        o Made mkdir deal properly with DOS slashes on OS/2 and Windows.

2006-11-24:
    - kBuild/bin/x86.win32/kmk.exe, kBuild/bin/x86.linux/kmk:
        o Rebuilt with current code. 
    - kBuild/footer.kmk:
        o Fixed some issues with SDKS and linking.
    - kmk:
        o Fixed bug in the sdk walking in kbuild_collect_source_prop.
        o Added --pretty-command-printing for simplifying makefile debugging.

2006-11-23:
    - dist/portage:
        o Hacked together two ebuilds for the Gentoo Portage.
    - Config.kmk, Makefile.kmk:
        o #6: Hacking *nix install build for Gentoo.
    - src/gmake/Makefile.kmk, src/lib/Makefile:
        o Always include header.kmk when testing BUILD_TARGET*.
    - kBuild/header.kmk:
        o BUILD_TYPE defaults to 'release'.
    - kBuild/tools/GCC*.kmk:
        o Cleaned up the -s and -g usage.
    - kBuild/bin/x86.win32, kBuild/bin/x86.linux:
        o Rebuilt with current code. 
    - kBuild/footer.kmk:
        o #3: Clean link output files.
    - kBuild/header.kmk, kmk:
        o #6: Did code changes for unix installation. Will do install goals when creating the gentoo ebuild.
        o #5: Proper versioning. (#5)
    - kmk:
        o Support SDK.$(BUILD_TARGET_ARCH) too (only bug).
        o Fixed a bug when gather SDKs.

2006-09-28:
    - kBuild/header.kmk, kBuild/env.sh, kBuild/tools/GCC4MACH.kmk, kBuild/bin/x86.darwin/, src/gmake, Config.kmk:
        o Initial Mac OS X / Darwin bootstrapping.

2006-09-23:
    - kBuild/footer.kmk:
        o Recursive template inheritance.
    - kBuild/header.kmk, kBuild/footer.kmk:
        o Removed some checks for features which are present in both gmake 3.81 and kmk.
          Anyone trying to bootstrap kBuild will have to build gmake 3.81 first.
    - Config.kmk, src/gmake, src/lib, src/kDepPre, src/kDepIDB:
        o Allow all kinds of ways of saying Windows in BUILD_TARGET.
    - kBuild/bin/x86.win32/:
        o Rebuilt kmk.exe.
    - kBuild/bin/x86.os2/:
        o Added kDepPre.exe and kDepIDB.exe.
        o Rebuilt kmk.exe and kmk_gmake.exe.
    - src/gmake:
        o Update KMK_FEATURES to include the optimizations from earlier this week.
        o Regenerated config.h.os2 and make adjustments to make it all
          build quietly on OS/2.
    - kBuild/envos2.cmd:
        o Environment script (REXX) for OS/2.

2006-09-19:
    - kBuild/bin/x86.win32/:
        o Rebuilt kmk.exe and kmk_gmake.exe.
    - src/gmake:
        o Fixed some optimization bugs in kbuild.c.

2006-09-18:
    - kBuild/gmake:
        o Don't send put j1 in MAKEFLAGS on windows, somehow that doesn't work.
        o Fixed some variable name typos in kbuild.c.
        o Fixed missing expansion in kbuild_lookup_variable.
        o A real attempt at fixing the double quoting problem on windows. ash seems to want a
          space between a double quote enclosing the last argument on the commandline and the
          double quote enclosing the commandline (the -c argument).
        o Fixed a newline escape problem when using batch_mode_shell on windows with kmk_ash.

2006-09-17:
    - kBuild/gmake:
        o Optimization summary: libc from ~21 seconds -> 7-8 seconds (os2/nt).
        o Optimized appending new stuff to variables. (major win)
        o Optimized variable memory value allocation avoiding a bunch of
          unnecessary copying and allocating.
        o Added kBuild specific functions for speeding up source processing.
        o Fixed assertion in w32_fixcase when shell doesn't have a full path
          and some other case.

2006-09-16:
    - kBuild/header.kmk:
        o Turn off command dependencies when using vanilla GNU Make.
    - kBuild/footer.kmk:
        o Made up my mind about the command depency blocker variables;
          NO_LINK_CMDS_DEPS and NO_COMPILE_CMDS_DEPS.
    - src/gmake:
        o Added kmk_gmake which is (almost) vanilla GNU Make.
        o kmk_ash wants batch files on windows or the double quotes will get screwed up.
        o Make sure kmk_builtin commands don't get batch files and odd weird stuff.
        o Cleaning up the modifications. Changes are now either configurable
          or marked, and dead stuff has been removed (dll shell).

2006-09-15:
    - kBuild/bin/x86.win32/kmk.exe:
        o Built new code.
    - kBuild/header.kmk:
        o DIRDEP is now the same on all platforms (fixed with 3.81 merge).
    - kBuild/bin/x86.linux/kmk:
        o Built new code on system360.
    - src/gmake:
        o Some parallel job fixes.
        o Regenerated config.h.linux on system360.
        o Fixed stat("dir/", &st) on windows and some warnings.
        o Merge with GNU Make v3.81 (vendor/gnumake/2005-05-16 -> vendor/gnumake/current).

2006-03-26:
    - kBuild/bin/x86.os2/:
        o Rebuilt the OS/2 tools.
    - src/gmake/variable.c, src/gmake/variable.h, src/gmake/make.h:
        o Attempt at speeding up performance on OS/2.
    - src/gmake/kmkbuiltin/rm.c:
        o OS/2 wants -R.
    - src/gmake/read.c:
        o Don't call glob() unless there is a wildcard pattern in the name.
          This speed things up a bit, at least on FreeBSD and especially on OS/2.
    - Config.kmk:
        o Added -pg for profile builds on OS/2.

2006-02-22: bird
    - kBuild/tools/MINGW32.kmk:
        o Corrected dependencies to match those generated by GCC3.

2006-02-09: bird

src/ash/bltin/echo.c
/*	$NetBSD: echo.c,v 1.12 2005/02/06 04:43:43 perry Exp $	*/

/*-
 * Copyright (c) 1991, 1993
 *	The Regents of the University of California.  All rights reserved.
 *
 * This code is derived from software contributed to Berkeley by
 * Kenneth Almquist.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 *
 *	@(#)echo.c	8.1 (Berkeley) 5/31/93
 */

/*
 * Echo command.
 *
 * echo is steeped in tradition - several of them!
 * netbsd has supported 'echo [-n | -e] args' in spite of -e not being
 * documented anywhere.
 * Posix requires that -n be supported, output from strings containing
 * \ is implementation defined
 * The Single Unix Spec requires that \ escapes be treated as if -e
 * were set, but that -n not be treated as an option.
 * (ksh supports 'echo [-eEn] args', but not -- so that it is actually
 * impossible to actually output '-n')
 *
 * It is suggested that 'printf "%b" "string"' be used to get \ sequences
 * expanded.  printf is now a builtin of netbsd's sh and csh.
 */

#define main echocmd

#include "bltin.h"

int
main(int argc, char **argv)
{
	char **ap;
	char *p;
	char c;
	int count;
	int nflag = 0;
	int eflag = 0;

	ap = argv;
	if (argc)
		ap++;

	if ((p = *ap) != NULL) {
		if (equal(p, "-n")) {
			nflag = 1;
			ap++;
		} else if (equal(p, "-e")) {
			eflag = 1;
			ap++;
		}
	}

	while ((p = *ap++) != NULL) {
		while ((c = *p++) != '\0') {
			if (c == '\\' && eflag) {
				switch (*p++) {
				case 'a':  c = '\a';  break;	/* bell */
				case 'b':  c = '\b';  break;
				case 'c':  return 0;		/* exit */
				case 'e':  c =  033;  break;	/* escape */
				case 'f':  c = '\f';  break;
				case 'n':  c = '\n';  break;
				case 'r':  c = '\r';  break;
				case 't':  c = '\t';  break;
				case 'v':  c = '\v';  break;
				case '\\':  break;		/* c = '\\' */
				case '0':
					c = 0;
					count = 3;
					while (--count >= 0 && (unsigned)(*p - '0') < 8)
						c = (c << 3) + (*p++ - '0');
					break;
				default:
					/* Output the '/' and char following */
					p--;
					break;
				}
			}
			putchar(c);
		}
		if (*ap)
			putchar(' ');
	}
	if (! nflag)
		putchar('\n');
	return 0;
}
